benchmarks/apps/sphng/sphng.py

```diff
@@ -0,0 +1,200 @@
+import os.path as path
+import reframe as rfm
+import reframe.utility.sanity as sn
+from benchmarks.modules.utils import SpackTest
+
+
+class SphngBase(SpackTest):
+
+    descr = 'Base class for Sphng'
+    valid_systems = ['*']
+    valid_prog_environs = ['*']
+    spack_spec = 'sphng@v1.0.0'
+    executable = 'sph_tree_rk_gradh'
+    time_limit = '0d0h10m0s'
+    exclusive_access = True
+
+    @run_before('compile')
+    def setup_build_system(self):
+        self.build_system.specs = [self.spack_spec]
+
+
+class SphngBase_ifile(SphngBase):
+
+    prerun_cmds = ['rm -rf fort* TEST* notify ifile sphng_setup.o']
+    executable_opts = ['initial ifile < ./setup.txt &> sphng_setup.o']
+
+    @sanity_function
+    def validate_ifile_generation(self):
+        return sn.assert_true(path.exists('ifile'))
+
+
+class SphngBase_evolution(SphngBase):
+
+    executable_opts = ['evolution ifile']
+
+    reference = {
+        'dial:slurm-local': {
+            'Total elapsed time': (10, None, None, 'minutes'),
+            },
+        }
+
+    @run_before('sanity')
+    def run_complete_pattern(self):
+        self.pattern = r'ended on'
+        self.sanity_patterns = sn.assert_found(self.pattern, 'test01')
+
+    @performance_function('minutes')
+    def get_elapsed_time(self):
+        return sn.extractsingle(r'cpu time used for this run :\s+(\S+)\s', 'test01', 1, float)
+
+    @run_before('performance')
+    def runtime_extract_pattern(self):
+        self.perf_variables = {
+                'Total elapsed time':self.get_elapsed_time()
+                }
+
+
+class Sphng_Single_Node_ifile(SphngBase_ifile):
+
+    mpi_tasks = parameter(2**i for i in range(0,8))
+
+    @run_after('setup')
+    def set_env_variables(self):
+
+        self.num_tasks = self.mpi_tasks
+
+        self.num_tasks_per_node = self.mpi_tasks
+
+        self.core_count_1_node = self.current_partition.processor.num_cpus // min(1, self.current_partition.processor.num_cpus_per_core)
+
+        self.thread_count = str(int(self.core_count_1_node/self.num_tasks_per_node))
+
+        self.descr = ('Running Sphng (Ifile) on ' + str(self.num_tasks_per_node) +' tasks and ' + \
+                       self.thread_count +  ' threads per node')
+
+        self.env_vars= {
+            'OMP_NUM_THREADS':self.thread_count,
+            'OMP_PLACES':'cores'
+        }
```

Comment:
> Looks overall good to me, I only left a couple of comments, but couldn't test myself. Thanks!

I will add you with read access to the github repos and then you should be able to run. Wouldn't hurt to get an external validation of these tests. In principle I don't see any problem adding anyone who wants to run the test. But perhaps a discussion we should have in the DiRAC meeting.

Reply:
> Here and in all other places below:
> 
> so that we don't completely [override the `env_vars` attribute](https://github.com/ukri-excalibur/excalibur-tests/pull/111/commits/94bacc1d67f9ad91f5cbf4caf4ce03b783a4ac1c).

Good spot. I will change these later today.